### PR TITLE
Feature/eks secondary ip

### PR DESCRIPTION
**Pull request by sreenivasmrpivot**
1. **What this PR does / why we need it**:
This PR adds secondary IP support for TMC EKS Cluster creation from Terraform

2. **Which issue(s) this PR fixes**

        (optional, in `fixes #<issue number>` format, will close the issue(s) when PR gets merged):

        Fixes #

3. **Additional information**

4. **Special notes for your reviewer**

### CODE REVIEW

**Review by asutoshpalai — CHANGES_REQUESTED** [review]
Please update the spec comparison helper in `helpers.go`. While you are at it, please also check at that all fields in the spec are also covered there. We might have missed some from previous MR.

Can you also add tests to `internal/resources/ekscluster/resource_ekscluster_test.go` and `ekscluster_flatten_test.go`?

File: internal/resources/ekscluster/resource_ekscluster.go [hunk 1]
```diff
@@ -271,6 +272,69 @@ var vpcSchema = &schema.Schema{
 	},
 }
 
+var addonsConfigSchema = &schema.Schema{
```
Comment (TaylorOno) [inline internal/resources/ekscluster/resource_ekscluster.go:275]: These should have descriptions

File: internal/models/ekscluster/addons_config.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
+/*
+Copyright 2022 VMware, Inc. All Rights Reserved.
```
Comment (TaylorOno) [inline internal/models/ekscluster/addons_config.go:2]: 2023

**Comment by TaylorOno** [comment]
Docs will need to be generated

File: examples/resources/ekscluster/cluster.tf [hunk 3]
```diff
@@ -49,6 +49,25 @@ resource "tanzu-mission-control_ekscluster" "tf_eks_cluster" {
           "subnet-06897e1063cc0cf4e",
         ]
       }
+
+      addons_config {
+        vpc_cni_config {
+          eni_configs {
+              eni_config {
```
[the remark this answers is not in the document]
Comment (sreenivasmrpivot) [reply examples/resources/ekscluster/cluster.tf:56]: @asutoshpalai structure of eni_configs and eni_config are different. eni_configs is not directly a collection or list of eni_config.

[the remark this answers is not in the document]
Comment (sreenivasmrpivot) [reply examples/resources/ekscluster/cluster.tf:56]: @asutoshpalai  I have made changes to helpers.go, added corresponding test. added tests to internal/resources/ekscluster/resource_ekscluster_test.go and ekscluster_flatten_test.go too.

I am not removing the eni_configs for now as I think this is just cosmetic at this point. This structure was auto generated using swagger and the API expects it in this specific structure.

File: internal/resources/ekscluster/resource_ekscluster.go (hunk 1 above)
Comment (sreenivasmrpivot) [reply internal/resources/ekscluster/resource_ekscluster.go:275]: @TaylorOno I have added the necessary descriptions

File: internal/models/ekscluster/addons_config.go (hunk 2 above)
Comment (sreenivasmrpivot) [reply internal/models/ekscluster/addons_config.go]: I have updated all new files generated with 2023

**Comment by sreenivasmrpivot** [comment]
> Docs will need to be generated

@TaylorOno I have generated the docs and pushed in new changes.

File: examples/resources/ekscluster/cluster.tf (hunk 3 above)
[the remark this answers is not in the document]
Comment (TaylorOno) [reply examples/resources/ekscluster/cluster.tf:56]: This just doesn't seem like the way Terraform likes to group data, I don't think Terraform should necessarily reflect how the API works

[the remark this answers is not in the document]
Comment (asutoshpalai) [reply examples/resources/ekscluster/cluster.tf:56]: Oh, we don't need to change the model generated by swagger files to make this change. This is a schema change. We can transform the data from schema to models in anyway we want while converting or flattening them.

File: docs/data-sources/package.md [hunk 4]
```diff
@@ -131,14 +131,14 @@ Read-Only:
 
 - `examples` (List of Object) (see [below for nested schema](#nestedobjatt--spec--values_schema--template--raw--examples))
 - `properties` (List of Object) (see [below for nested schema](#nestedobjatt--spec--values_schema--template--raw--properties))
-- `title` (String) Title of object.
+- `title` (String)
```
Comment (asutoshpalai) [inline docs/data-sources/package.md:134]: I am guessing these are auto-generated changes, but can you please exclude them from this MR?

File: internal/resources/ekscluster/ekscluster_flatten_test.go [hunk 5]
```diff
@@ -15,20 +15,170 @@ import (
 
 func TestFlattenCluterSpec(t *testing.T) {
 	tests := []struct {
-		description string
-		getInput    func() (*eksmodel.VmwareTanzuManageV1alpha1EksclusterSpec, []*eksmodel.VmwareTanzuManageV1alpha1EksclusterNodepoolDefinition)
-		expected    []interface{}
+		description  string
+		getInput     func(addonsConfig *eksmodel.VmwareTanzuManageV1alpha1EksclusterAddonsConfig) (*eksmodel.VmwareTanzuManageV1alpha1EksclusterSpec, []*eksmodel.VmwareTanzuManageV1alpha1EksclusterNodepoolDefinition)
+		addonsConfig *eksmodel.VmwareTanzuManageV1alpha1EksclusterAddonsConfig
```
Comment (asutoshpalai) [inline internal/resources/ekscluster/ekscluster_flatten_test.go:20]: `getInput` is a generic function which can be used to add (or remove) `addonsConfig`. We don't need another field to specify the `addonsConfig` here.

File: internal/resources/ekscluster/ekscluster_flatten_test.go [hunk 6]
```diff
@@ -472,20 +649,24 @@ func TestFlattenConfig(t *testing.T) {
...
-			ProxyName:        "test-prooxy",
-			Config:           getConfig(),
-		}, []*eksmodel.VmwareTanzuManageV1alpha1EksclusterNodepoolDefinition{
-			{
-				Info: &eksmodel.VmwareTanzuManageV1alpha1EksclusterNodepoolInfo{
-					Description: "test np",
-					Name:        "test-np",
-				},
-				Spec: getNodepoolSpec(),
+func getClusterSpec(addonsConfig *eksmodel.VmwareTanzuManageV1alpha1EksclusterAddonsConfig) (*eksmodel.VmwareTanzuManageV1alpha1EksclusterSpec, []*eksmodel.VmwareTanzuManageV1alpha1EksclusterNodepoolDefinition) {
```
Comment (asutoshpalai) [inline internal/resources/ekscluster/ekscluster_flatten_test.go:652]: Let this method return a full cluster spec. For the cases where we don't need the addon config, we can overwrite it in the `getInput` function as done for the "empty config" test case.

File: internal/resources/ekscluster/helpers_test.go [hunk 7]
```diff
@@ -582,12 +582,120 @@ func TestClusterSpecEqual(t *testing.T) {
 			},
 			result: true,
 		},
+		{
+			name: "addon configs are set equal",
```
Comment (asutoshpalai) [inline internal/resources/ekscluster/helpers_test.go:586]: I don't see a set equality test case. They both look identically ordered to me

**Review by asutoshpalai — CHANGES_REQUESTED** [review]

File: examples/resources/ekscluster/cluster.tf (hunk 3 above)
[the remark this answers is not in the document]
Comment (sreenivasmrpivot) [reply examples/resources/ekscluster/cluster.tf]: @asutoshpalai  I have addressed the structure changes mentioned above.

File: internal/resources/ekscluster/ekscluster_flatten_test.go (hunk 5 above)
Comment (sreenivasmrpivot) [reply internal/resources/ekscluster/ekscluster_flatten_test.go]: @asutoshpalai  I have made changes as requested

File: internal/resources/ekscluster/ekscluster_flatten_test.go (hunk 6 above)
Comment (sreenivasmrpivot) [reply internal/resources/ekscluster/ekscluster_flatten_test.go]: @asutoshpalai I have made changes as requested

File: docs/data-sources/package.md (hunk 4 above)
Comment (sreenivasmrpivot) [reply docs/data-sources/package.md:134]: @asutoshpalai  These were auto generated with tfplugindocs as per @TaylorOno  request and I have generated docs few times across pushes. It will be very difficult to separate from the MR now.

File: internal/resources/ekscluster/helpers_test.go (hunk 7 above)
Comment (sreenivasmrpivot) [reply internal/resources/ekscluster/helpers_test.go:586]: @asutoshpalai  equality test works now and I have pushed the changes.

**Review by asutoshpalai — DISMISSED** [review]

**Review by asutoshpalai — DISMISSED** [review]

File: docs/resources/ekscluster.md [hunk 8]
```diff
@@ -260,6 +278,34 @@ Optional:
 - `security_groups` (Set of String) Security groups for the cluster VMs
 
 
+<a id="nestedblock--spec--config--addons_config"></a>
+### Nested Schema for `spec.config.addons_config`
+
+Optional:
+
+- `vpc_cni_config` (Block List, Max: 1) VPC CNI addon config contains the configuration for the VPC CNI addon of the cluster (see [below for nested schema](#nestedblock--spec--config--addons_config--vpc_cni_config))
```
Comment (DheerajSShetty) [inline docs/resources/ekscluster.md:286]: Lets add more details about vp cni config, eni config, the subnets, AZs and security groups in ENiConfigs, explaining the dependency of the AZ here with the AZs for Nodepools

File: examples/resources/ekscluster/cluster.tf [hunk 9]
```diff
@@ -49,6 +49,23 @@ resource "tanzu-mission-control_ekscluster" "tf_eks_cluster" {
           "subnet-06897e1063cc0cf4e",
         ]
       }
+
+      addons_config {
+        vpc_cni_config {
+          eni_config {
+            id = "subnet-0a680171b6330619f" // Required, should belong to the same VPC as the cluster
```
Comment (DheerajSShetty) [inline examples/resources/ekscluster/cluster.tf:56]: would be a good place to talk about AZ as well

File: examples/resources/ekscluster/cluster.tf [hunk 10]
```diff
@@ -49,6 +49,23 @@ resource "tanzu-mission-control_ekscluster" "tf_eks_cluster" {
...
+            security_groups = [
+              "sg-00c96ad9d02a22522",
```
Comment (DheerajSShetty) [inline examples/resources/ekscluster/cluster.tf:58]: Example should include an empty security group list as well

File: internal/models/ekscluster/vpccni_addon_config.go [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+import (
+	"github.com/go-openapi/swag"
+)
+
+// VmwareTanzuManageV1alpha1EksclusterVpcCniAddonConfig EKS vpc-cni addon configuration
+//
+// swagger:model vmware.tanzu.manage.v1alpha1.ekscluster.VpcCniAddonConfig
+type VmwareTanzuManageV1alpha1EksclusterVpcCniAddonConfig struct {
+
+	// Enable the Kubernetes vpc-cni addon.
```
Comment (DheerajSShetty) [inline internal/models/ekscluster/vpccni_addon_config.go:17]: VP CNI is enabled by default, this comment should be fixed

File: internal/resources/ekscluster/ekscluster_flatten_test.go [hunk 12]
```diff
@@ -73,6 +208,30 @@ func TestFlattenCluterSpec(t *testing.T) {
 									},
 								},
 							},
+							"addons_config": []interface{}{
+								map[string]interface{}{
```
Comment (DheerajSShetty) [inline internal/resources/ekscluster/ekscluster_flatten_test.go:212]: Lets add a test for empty security group, and we need to make sure if an empty security group is specified, the TF state file has the cluster security group set after cluster creation.

File: docs/resources/ekscluster.md (hunk 8 above)
Comment (sreenivasmrpivot) [reply docs/resources/ekscluster.md]: @DheerajSShetty  fixed this.

File: examples/resources/ekscluster/cluster.tf (hunk 9 above)
Comment (sreenivasmrpivot) [reply examples/resources/ekscluster/cluster.tf]: @DheerajSShetty  fixed this.

File: examples/resources/ekscluster/cluster.tf (hunk 10 above)
Comment (sreenivasmrpivot) [reply examples/resources/ekscluster/cluster.tf:58]: @DheerajSShetty  fixed this.

File: internal/models/ekscluster/vpccni_addon_config.go (hunk 11 above)
Comment (sreenivasmrpivot) [reply internal/models/ekscluster/vpccni_addon_config.go]: @DheerajSShetty  fixed this.

File: internal/resources/ekscluster/ekscluster_flatten_test.go (hunk 12 above)
Comment (sreenivasmrpivot) [reply internal/resources/ekscluster/ekscluster_flatten_test.go:212]: @DheerajSShetty  the data comes back from API and adding a empty security group would not serve the purpose as the expectation is for the API to return a security group back from the cluster though it was not added in the eniconfig. This is in line with our discussion. I have manually tested the terraform file with empty security groups under the eni_config and it still sets the cluster security group under eni_config as expected in the state file.

File: docs/resources/ekscluster.md [hunk 13]
```diff
@@ -283,25 +283,25 @@ Optional:
...
+- `vpc_cni_config` (Block List, Max: 1) VPC CNI addon config contains the configuration for the VPC CNI addon of the cluster. (see [below for nested schema](#nestedblock--spec--config--addons_config--vpc_cni_config))
 
 <a id="nestedblock--spec--config--addons_config--vpc_cni_config"></a>
 ### Nested Schema for `spec.config.addons_config.vpc_cni_config`
 
 Optional:
 
-- `eni_config` (Block List) ENI config for the VPC CNI addon (see [below for nested schema](#nestedblock--spec--config--addons_config--vpc_cni_config--eni_config))
+- `eni_config` (Block List) ENI config is the VPC CNI Elastic Network Interface config for providing the configuration of subnet and security groups for pods in each AZ.  with subnet as a key and list of security group as value. The subnets are expected to be in different AZs. Nodepools will be created only in the az where the pod subnets resides. (see [below for nested schema](#nestedblock--spec--config--addons_config--vpc_cni_config--eni_config))
```
Comment (DheerajSShetty) [inline docs/resources/ekscluster.md:293]: `with subnet as a key and list of security group as value.`
Is Subnet a key or just a member of ENIConfig?
`The subnets are expected to be in different AZs. Nodepools will be created only in the az where the pod subnets resides.`
Let's not use the term pod subnet as we don't use this term anywhere else. Does this sound better? `Nodepool subnets need to be in the same AZ as the AZs used in ENIConfig`

File: examples/resources/ekscluster/cluster.tf [hunk 14]
```diff
@@ -50,19 +50,16 @@ resource "tanzu-mission-control_ekscluster" "tf_eks_cluster" {
         ]
       }
 
-      addons_config {
+      addons_config { // this whole section is optional
         vpc_cni_config {
           eni_config {
-            id = "subnet-0a680171b6330619f" // Required, should belong to the same VPC as the cluster
-            security_groups = [
+            id = "subnet-0a680171b6330619f" // Required, should belong to the same VPC as the cluster, subnets provided in vpc_cni_config are expected to be in different AZs
```
Comment (DheerajSShetty) [inline examples/resources/ekscluster/cluster.tf:56]: It need not be in the same VPC as the Cluster. subnets provided in list of end_config within vpc_cni_config are expected to be in different AZs

File: examples/resources/ekscluster/cluster.tf [hunk 15]
```diff
@@ -50,19 +50,16 @@ resource "tanzu-mission-control_ekscluster" "tf_eks_cluster" {
...
+            security_groups = [ //optional
               "sg-00c96ad9d02a22522",
             ]
           }
           eni_config {
```
Comment (DheerajSShetty) [inline examples/resources/ekscluster/cluster.tf:61]: good to point out that the EKS cluster security group will be used here

File: internal/models/ekscluster/vpccni_addon_config.go [hunk 16]
```diff
@@ -14,7 +14,7 @@ import (
 // swagger:model vmware.tanzu.manage.v1alpha1.ekscluster.VpcCniAddonConfig
 type VmwareTanzuManageV1alpha1EksclusterVpcCniAddonConfig struct {
 
-	// Enable the Kubernetes vpc-cni addon.
+	// Additional configuration settings for vpc-cni addon.
```
Comment (DheerajSShetty) [inline internal/models/ekscluster/vpccni_addon_config.go:17]: Can we write about ENIConfigs here?

File: docs/resources/ekscluster.md (hunk 13 above)
Comment (sreenivasmrpivot) [reply docs/resources/ekscluster.md]: @DheerajSShetty fixed this

File: examples/resources/ekscluster/cluster.tf (hunk 14 above)
Comment (sreenivasmrpivot) [reply examples/resources/ekscluster/cluster.tf]: @DheerajSShetty fixed this

File: examples/resources/ekscluster/cluster.tf (hunk 15 above)
Comment (sreenivasmrpivot) [reply examples/resources/ekscluster/cluster.tf:61]: @DheerajSShetty fixed this

File: internal/models/ekscluster/vpccni_addon_config.go (hunk 16 above)
Comment (sreenivasmrpivot) [reply internal/models/ekscluster/vpccni_addon_config.go]: @DheerajSShetty fixed this

**Comment by DheerajSShetty** [comment]
LGTM

**Review by DheerajSShetty — APPROVED** [review]

**Review by asutoshpalai — DISMISSED** [review]

**Review by DheerajSShetty — APPROVED** [review]
